nodego/supervisor.go

```diff
@@ -0,0 +1,255 @@
+	}
+
+	if loggingContext.currentLogBatch == nil {
+		return os.Stderr.Write(entry.consoleOutput(loggingContext.functionExecutionID))
+	}
+
+	// Start a new batch if the current one would grow too much.
+	if len(loggingContext.currentLogBatch.Entries)+1 > maxLogBatchEntries || loggingContext.currentLogBatch.payloadLength+len(entry.TextPayload) > maxLogBatchLength {
+		startNewLogBatch()
+	}
+
+	loggingContext.currentLogBatch.addEntry(entry)
+	return len(entry.TextPayload), nil
+}
+
+func (batch *logBatch) report() error {
+	if len(batch.Entries) == 0 {
+		return nil
+	}
+
+	if err := postToSupervisor("/_ah/log", batch, supervisorLogTimeout); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func flushLogBatches() {
+	select {
+	case logBatch := <-logBatchesToReport:
+		go flushLogBatches()
+		if err := logBatch.report(); err != nil {
+			fmt.Fprintln(os.Stderr, err.Error())
+			killInstance()
+		}
+	default:
+		startNewLogBatch()
+	}
+}
+
+func logBatchReportWorker() {
+	for logBatch := range logBatchesToReport {
+		// TODO it may be nice to have a time base buffering method on top of this.
+		<-logBatch.ready
+
+		loggingContext.Lock()
+		if logBatch == loggingContext.currentLogBatch {
+			startNewLogBatch()
+		}
+		loggingContext.Unlock()
+
+		if err := logBatch.report(); err != nil {
+			fmt.Fprintln(os.Stderr, err.Error())
+			killInstance()
+		}
+	}
+}
+
+func newSupervisorRequest(path string, v interface{}) (*http.Request, error) {
+	postData, err := json.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", (&url.URL{
+		Scheme: "http",
+		Host:   supervisorHostname + ":" + supervisorInternalPort,
+		Path:   path,
+	}).String(), bytes.NewBuffer(postData))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Content-Length", strconv.Itoa(len(postData)))
+
+	return req, nil
+}
+
+func postToSupervisor(path string, v interface{}, timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	req, err := newSupervisorRequest(path, v)
+	if err != nil {
+		return err
+	}
+
+	resp, err := ctxhttp.Do(ctx, nil, req)
+	if err == ctx.Err() {
+		return errors.New("timeout when calling supervisor")
+	} else if err != nil {
+		return fmt.Errorf("error when calling supervisor: %s\n\n%s\n", err.Error(), debug.Stack())
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		err = fmt.Errorf("incorrect response code from supervisor: %d\n", resp.StatusCode)
+	}
+
+	return err
+}
+
+func killInstance() {
+	postToSupervisor("/_ah/kill", nil, supervisorKillTimeout)
+	os.Exit(16)
+}
+
+func loggerMiddleware(handler http.Handler) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		loggingContext.Lock()
+		loggingContext.functionExecutionID = r.Header.Get("Function-Execution-Id")
+		loggingContext.Unlock()
+
+		defer func() {
+			go flushLogBatches()
```

Comment:
Why not just call startNewLogBatch? There is already a goroutine waiting that will flush as soon as it can. Plus, that function is already called synchronously in handler execution elsewhere.

Reply:
This is what I was trying to say before. In fact, there doesn't even need to be a `startNewLogBatch` call because the goroutine will create a new batch if the last one is not empty. `flushLogBatches` actually sends all the remaining log batches asyncronously.

Reply:
`startNewLogBatch` will kick off the flush the next time the `logBatchReportWorker` goroutine is scheduled. Starting a new goroutine won't help as there is no guarantee that it will be scheduled before the already waiting `logBatchReportWorker` goroutine.

Reply:
If there is an empty batch at the end of the queue, then calling `startNewLogBatch` will result in two empty batches at the end. When `logBatchReportWorker` finds an empty batch, it will wait until something is added to it, and since new logs are only added to the last batch, it will be waiting forever.

`flushLogBatches` will go through all remaining batches regardless of whether they are empty are not and finally creating a new empty batch when the queue is dry.

There does not need to be a guarentee that it runs before a `logBatchReportWorker` loop. If it runs, afterwards and there are no more batches, it will just won't do anything. If it runs before and takes the last batch, `logBatchReportWorker` will wait until `flushLogBatches` creates a new batch and will just wait until that new batch has log added to it.